src/support.h

```diff
@@ -124,6 +134,23 @@ inline static const v8::MaybeLocal<T> ptr_to_maybe_local(const T* ptr) {
   return *reinterpret_cast<const v8::MaybeLocal<T>*>(&ptr);
 }
 
+template <class T>
+inline static const T* maybe_to_ptr(v8::Maybe<T> maybe) {
+  if (maybe.IsNothing()) {
+    return nullptr;
+  }
+  return maybe.FromJust();
+}
+
+template <class T>
+inline static const v8::Maybe<T> ptr_to_maybe(const T* ptr) {
+  static_assert(sizeof(v8::Maybe<T>) == sizeof(T*), "");
```

Comment:
Does this pass? The non-specialized definition of `Maybe` is:

```c++
template <class T>
class Maybe {
  bool has_value_;
  T value_; 
}
```

... and I could not find a specialization for pointers.

I think you could achieve what you need by using the `Maybe<T> Maybe::Just(const T& t);` method. 
Like `return Maybe::Just(*ptr);`.

Reply:
Additionally, the assertion is incorrect. It should be:

```
static_assert(sizeof(v8::Maybe<T*>) == sizeof(T*), "");
```

Reply:
Actually, shouldn't Maybe be mapped to a Rust Option instead of a simple pointer?

Reply:
> I think `MaybeLocal` can be directly mapped to `Option<Box>` because both can be represented as a nullable pointer. But for Maybe to be mapped to Option; an additional boolean is required I think.

What matters here is that the actual layout of the C++ and Rust types need match, and they also need to be *guaranteed* to match (otherwise clang / rust will complain that the type is not FFI-safe).

Indeed, `Option<NonNull<T>>` and `Option<Box<T>>` have the same layout as a raw pointer (`*mut T`) and rust *does* explicitly guarantee this -- but this is the exception, most non-primitive types are not considered FFI-safe.

`Maybe<T>`  probably needs this rust definition:
```
[repr(C)]
struct Maybe<T> {
  has_value: bool;
  value: T
}
```

The above applies to the FFI interface. If we can avoid it we would not expose `Maybe<T>` to the end user, but rather use the applicable idiomatic rust type, e.g.  `Option<&T>` in the end-user interface.